pkg/controller/dns/external_dns.go

```diff
@@ -0,0 +1,167 @@
+	}
+
+	retriever = retriever.Remove(common.RetrieverFromGk(
+		nil, // our compare strat is ignore labels
+		schema.GroupKind{
+			Group: corev1.GroupName,
+			Kind:  "Namespace",
+		}),
+		common.RemoveOpt{
+			CompareStrat: common.IgnoreLabels, // ignore labels, we never want to clean namespaces
+		})
+
+	return common.NewCleaner(manager, "externalDnsCleaner", retriever)
+}
+
+// NewExternalDns starts all resources required for external dns
+func NewExternalDns(manager ctrl.Manager, conf *config.Config, self *appsv1.Deployment) error {
+	instances := instances(conf, self)
+
+	deployInstances := filterAction(instances, deploy)
+	deployRes := getResources(deployInstances)
+	if err := addExternalDnsReconciler(manager, deployRes); err != nil {
+		return err
+	}
+
+	cleanObjs := cleanObjs(instances)
+	if err := addExternalDnsCleaner(manager, cleanObjs); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func instances(conf *config.Config, self *appsv1.Deployment) []instance {
+	// public
+	publicConfig := publicConfig(conf)
+	publicAction := deploy
+	if len(conf.PublicZoneConfig.ZoneIds) == 0 {
+		publicAction = clean
+	}
+	publicResources := manifests.ExternalDnsResources(conf, self, []*manifests.ExternalDnsConfig{publicConfig})
+
+	// private
+	privateConfig := privateConfig(conf)
+	privateAction := deploy
+	if len(conf.PrivateZoneConfig.ZoneIds) == 0 {
+		privateAction = clean
+	}
+	privateResources := manifests.ExternalDnsResources(conf, self, []*manifests.ExternalDnsConfig{privateConfig})
+
+	return []instance{
+		{
+			config:    publicConfig,
+			resources: publicResources,
+			action:    publicAction,
+		},
+		{
+			config:    privateConfig,
+			resources: privateResources,
+			action:    privateAction,
+		},
+	}
+}
+
+func publicConfig(conf *config.Config) *manifests.ExternalDnsConfig {
+	return &manifests.ExternalDnsConfig{
+		TenantId:           conf.TenantID,
+		Subscription:       conf.PublicZoneConfig.Subscription,
+		ResourceGroup:      conf.PublicZoneConfig.ResourceGroup,
+		Provider:           manifests.PublicProvider,
+		DnsZoneResourceIDs: conf.PublicZoneConfig.ZoneIds,
+	}
+}
+
+func privateConfig(conf *config.Config) *manifests.ExternalDnsConfig {
+	return &manifests.ExternalDnsConfig{
+		TenantId:           conf.TenantID,
+		Subscription:       conf.PrivateZoneConfig.Subscription,
+		ResourceGroup:      conf.PrivateZoneConfig.ResourceGroup,
+		Provider:           manifests.PrivateProvider,
+		DnsZoneResourceIDs: conf.PrivateZoneConfig.ZoneIds,
+	}
+}
+
+func filterAction(instances []instance, action action) []instance {
+	var ret []instance
+	for _, i := range instances {
+		if i.action == action {
+			ret = append(ret, i)
+		}
+	}
+
+	return ret
+}
+
+func getResources(instances []instance) []client.Object {
+	var ret []client.Object
+	for _, i := range instances {
+		ret = append(ret, i.resources...)
+	}
+	return ret
+}
+
+func getLabels(instances ...instance) map[string]string {
```

Comment:
love the labels approach here super clean

Reply:
Thanks :)!